### PR TITLE
My Jetpack: tidy Product card component

### PR DESCRIPTION
**Pull request by retrofox**
This PR iterates over ProductCard components in order to simplify and reduce the code.  There is a bunch of repeated code that can be simplified thanks to the `useProduct()` hook, name conventions, etc. The goal is, again, to make the code more eligible, simpler, but without removing the ability to customize some of them when it's required. `showDeactivate` is an example of it.

image

Fixes #

#### Changes proposed in this Pull Request:

* Introduces <ConnectedProductCard/> component.
* SImplifies code in product cards by using <ConnectedProductCard/> instead of <ProductCard/> directly

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:

* Go to My Jetpack dashboard. Confirm everything works as expected:
  *  Activate/Deactivate Product
  * Add Product link leads to the Interstitial page
  * Confirm it triggers events
  * etc

### CODE REVIEW

**Comment by oskosk** [comment]
Can you please clarify what are the changes introduced here under `Changes proposed in this Pull Request:`. The diff is huge

File: projects/packages/my-jetpack/_inc/components/product-card/index.jsx [hunk 1]
```diff
@@ -183,25 +179,34 @@ const ProductCard = props => {
 		recordEvent( 'jetpack_myjetpack_product_card_fixconnection_click', {
 			product: slug,
 		} );
-		onFixConnection();
-	}, [ slug, onFixConnection, recordEvent ] );
+		navigateToConnectionPage();
+	}, [ slug, navigateToConnectionPage, recordEvent ] );
+
+	const ActionButtonInstance = actionButtonProps => (
```
Comment (oskosk) [inline projects/packages/my-jetpack/_inc/components/product-card/index.jsx:185]: Let's avoid things like this `ActionButtonInstance` generate. It doesn't categarize as elegant to me (which seems to be the purpose of this PR). Component names should be clear in terms of what they represent. functions don't have instances :).

And let's avoid it especially if it's only purpose is to enable the anti-pattern `{ ...actionButtonProps }`

File: projects/packages/my-jetpack/_inc/components/product-card/index.jsx [hunk 2]
```diff
@@ -183,25 +179,34 @@ const ProductCard = props => {
...
+		<ActionButton
+			status={ status }
+			isFetching={ isFetching }
+			name={ name }
+			admin={ admin }
+			onActivate={ activateHandler }
+			onFixConnection={ fixConnectionHandler }
+			onManage={ manageHandler }
+			{ ...actionButtonProps }
```
Comment (oskosk) [inline projects/packages/my-jetpack/_inc/components/product-card/index.jsx:194]: This is an anti-patterns. It hides things... We're refactoring a whole bunch of code so to not have to copy/paste props in the future. It's better to be clear and delcarative than to hide the properties to the readers of the code


* https://vhudyma-blog.eu/react-antipatterns-props-spreading/
* https://codeburst.io/react-anti-pattern-jsx-spread-attributes-59d1dd53677f

**Review by oskosk — CHANGES_REQUESTED** [review]
I can't really share the sensation that the changes here make the code clearer... :( Left a comment asking for clarification on the changes in the PR descrption.

**Comment by oskosk** [comment]
I think this PR is going against @renatoagds ’s intention of keeping  productcard being a presentational component.

I just realized that if we merge this PR then the card become super obscure as the properties that update its behaviour are
no longer visible on the react dev tools

**Comment by retrofox** [comment]
> I think this PR is going against @renatoagds ’s intention of keeping productcard being a presentational component.

Yeah, that's ok and I'm aware of that.

> I just realized that if we merge this PR then the card become super obscure as the properties that update its behaviour are no longer visible on the react dev tools

I disagree, quite the opposite from my PoV. Anyway, it isn't a big deal, either.
If you take a look at the code, basically we are repeating the same code in eight files. Every time we need to add something to the product card, we need to copy and paste the same code over all files, which doesn't make sense, again, IMO.

**Comment by retrofox** [comment]
thanks for your review

**Comment by oskosk** [comment]
>If you take a look at the code, basically we are repeating the same code in eight files. Every time we need to add something to the product card, we need to copy and paste the same code over all files, which doesn't make sense, again, IMO.

The fact that the implementation here was not elegant does not invalidat your point. You could have created a components called `<ConnectedProductCard slug="anti-spam" />` that would have removed the same code from eight files.

**Comment by oskosk** [comment]
Reopend and pushed three commits

* Introduce ConnectedProductCard component 0301506c138838e87c0c4f93cbed83aa55d1f851
* Roll back changes to ProductCard 74e678885017182ee50f6ae67863425041556d64 
* Make the specific product cards use the ConnectedProductCard component instead of ProductCard directly. 25b8722b738317a7eb9aecdfba24772f4326620b

@retrofox please let me know if this version accomplishes the same intention you wanted to achieve. (reducing the boilerplate on the specific product cards).

File: projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx [hunk 3]
```diff
@@ -1,38 +1,21 @@
...
+import React from 'react';
 import PropTypes from 'prop-types';
 
 /**
  * Internal dependencies
  */
-import ProductCard from '../product-card';
-import { useProduct } from '../../hooks/use-product';
-import { AntiSpamIcon } from '../icons';
+import ProductCard from './connected-product-card';
```
Comment (renatoagds) [inline projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx:10]: I just have minimal concern about using the component name different than the filename, but not a big problem, only a minimal point in terms of readability.

File: projects/packages/my-jetpack/_inc/components/product-cards-section/connected-product-card.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	const onManage = useCallback( () => {
+		window.location = manageUrl;
+	}, [ manageUrl ] );
+
+	return (
+		<ProductCard
+			name={ name }
+			description={ description }
+			status={ status }
+			icon={ <AntiSpamIcon /> }
```
Comment (renatoagds) [inline projects/packages/my-jetpack/_inc/components/product-cards-section/connected-product-card.jsx:34]: We should receive the icon, or map it based on the slug (I prefer the first one, but is up to you), this way we'll show `AntiSpamIcon` in all cards.

**Review by renatoagds — CHANGES_REQUESTED** [review]
The icon point is the required one.

File: projects/packages/my-jetpack/_inc/components/product-cards-section/connected-product-card.jsx (hunk 4 above)
Comment (retrofox) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/connected-product-card.jsx:34]: fixing...

Comment (retrofox) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/connected-product-card.jsx]: before | after
----|-----
image | image

> We should receive the icon, or map it based on the slug (I prefer the first one, but is up to you), 

well... the spirit of this new _connected_ component is to resolve these handling magically, in the opposite direction than the ProductCard component.

[Addressed here](https://github.com/Automattic/jetpack/pull/23224/commits/b8ab3bf9a7b5e19c1730230f1d5a1ed0cf7b241d)

**Comment by retrofox** [comment]
> The icon point is the required one.

[Addressed here](https://github.com/Automattic/jetpack/pull/23224/commits/b8ab3bf9a7b5e19c1730230f1d5a1ed0cf7b241d)

File: projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx (hunk 3 above)
Comment (retrofox) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx:10]: do you mean creating a new folder for this?

Comment (renatoagds) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx:10]: No, I mean:

```
import ConnectedProductCard from './connected-product-card`;
```

**Review by renatoagds — APPROVED** [review]
LGTM

File: projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx (hunk 3 above)
Comment (oskosk) [reply projects/packages/my-jetpack/_inc/components/product-cards-section/anti-spam-card.jsx:10]: He means he would be more comfortable if it read 
```
import ConnectedProductCard from './connected-product-card';
```

**Review by oskosk — APPROVED** [review]
LGTM!

**Pull request merged**